python/ray/tune/trainable/trainable.py

```diff
@@ -569,41 +570,71 @@ def _get_latest_remote_available_checkpoint(self) -> Optional[str]:
 
         return max(checkpoint_candidates)
 
-    def _maybe_save_to_cloud(self, checkpoint_dir: str) -> bool:
+    def _maybe_save_artifacts_to_cloud(self) -> bool:
+        if not self.sync_config.sync_artifacts:
+            return False
+
+        if self._last_artifact_sync_iter == self.iteration:
+            # No need to sync again, if we have already synced this iteration.
+            return False
+        self._last_artifact_sync_iter = self.iteration
+        with warn_if_slow(
+            name="trial_artifact_cloud_upload",
+            message=(
+                "Uploading trial artifacts took {duration:.3f} s, which may be a "
+                "performance bottleneck. Consider saving fewer/smaller artifacts to "
+                "the trial log directory, or disable artifact syncing with "
+                "`SyncConfig(sync_artifacts=False)`."
+            ),
+            # Log a warning if upload time surpasses 10s
+            threshold=10,
+            disable=not self.uses_cloud_checkpointing,
+        ):
+            # Avoid double uploading checkpoints and driver artifacts,
+            # if those live in the same directory
+            exclude = ("checkpoint_*",) + EXPR_FILES
+            uploaded = self._maybe_save_to_cloud(self.logdir, exclude=exclude)
+        return uploaded
+
+    def _maybe_save_to_cloud(self, local_dir: str, exclude: List[str] = None) -> bool:
+        """Saves the given directory to the cloud. This is used for checkpoint
+        and artifact uploads to cloud.
+
+        Args:
+            local_dir: The local directory to upload to the `remote_checkpoint_dir`
+
+        Returns:
+            bool: True if (successfully) saved to cloud
+        """
         if not self.uses_cloud_checkpointing:
             return False
 
-        if self.custom_syncer:
-            self.custom_syncer.sync_up(
-                checkpoint_dir, self._storage_path(checkpoint_dir)
-            )
-            self.custom_syncer.wait_or_retry(
+        syncer = self.sync_config.syncer
+        assert syncer
+
+        checkpoint_uri = self._storage_path(local_dir)
+        _ensure_directory(checkpoint_uri)
```

Comment:
I needed to add this in for tests to pass. Otherwise, writing to the local fs complains that the checkpoint directory doesn't exist. Is this a problem for generic filesys?

Reply:
It should be fine - let's check this in the tune cloud tests!